packages/cli/command_help.pony

```diff
@@ -0,0 +1,196 @@
+  let spec: CommandSpec box
+
+  new _create(parent': (CommandHelp box | None), spec': CommandSpec box) =>
+    parent = parent'
+    spec = spec'
+
+  fun fullname(): String =>
+    match parent
+    | let p: CommandHelp => p.fullname() + " " + spec.name
+    else
+      spec.name
+    end
+
+  fun string(): String => fullname()
+
+  fun help_string(): String =>
+    let w: StringWriter ref = StringWriter
+    print_help(w)
+    w.string()
+
+  fun print_help(w: Writer) =>
+    print_usage(w)
+    w.write("\n")
+    w.write(spec.descr + "\n")
+    w.write("\n")
+
+    let options = all_options()
+    if options.size() > 0 then
+      w.write("Options:\n")
+      print_options(w, options)
+    end
+    if spec.commands.size() > 0 then
+      w.write("\nCommands:\n")
+      print_commands(w)
+    end
+    let args = spec.args
+    if args.size() > 0 then
+      w.write("\nArgs:\n")
+      print_args(w, args)
+    end
+
+  fun print_usage(w: Writer) =>
+    w.write("usage: " + fullname())
+    if any_options() then
+      w.write(" [<options>]")
+    end
+    if spec.commands.size() > 0 then
+      w.write(" <command>")
+    end
+    if spec.args.size() > 0 then
+      for a in spec.args.values() do
+        w.write(" " + a.help_string())
+      end
+    else
+      w.write(" [<args> ...]")
+    end
+    w.write("\n")
+
+  fun print_options(w: Writer, options: Array[OptionSpec box] box) =>
+    let cols = Array[(String,String)]()
+    for o in options.values() do
+      cols.push(("  " + o.help_string(), o.descr))
+    end
+    Columns.print(w, cols)
+
+  fun print_commands(w: Writer) =>
+    let cols = Array[(String,String)]()
+    _list_commands(spec, cols, 1)
+    Columns.print(w, cols)
+
+  fun _list_commands(cs: CommandSpec box, cols: Array[(String,String)], level: USize) =>
+    for c in cs.commands.values() do
+      cols.push((Columns.indent(level*2) + c.help_string(), c.descr))
+      _list_commands(c, cols, level + 1)
+    end
+
+  fun print_args(w: Writer, args: Array[ArgSpec] box) =>
+    let cols = Array[(String,String)]()
+    for a in args.values() do
+      cols.push(("  " + a.help_string(), a.descr))
+    end
+    Columns.print(w, cols)
+
+  fun any_options(): Bool =>
+    if spec.options.size() > 0 then
+      true
+    else
+      match parent
+      | let p: CommandHelp => p.any_options()
+      else
+        false
+      end
+    end
+
+  fun all_options(): Array[OptionSpec box] =>
+    let options = Array[OptionSpec box]()
+    _all_options(options)
+    options
+
+  fun _all_options(options: Array[OptionSpec box]) =>
+    match parent
+    | let p: CommandHelp => p._all_options(options)
+    end
+    for o in spec.options.values() do
+      options.push(o)
+    end
+
+"""
+This interface and two classes allow the help output to go to a general Writer
+that can be implemented for string creation or to an OutStream. Or other
+targets. This could be polished up and moved into a more central package.
+"""
```

Comment:
I'd prefer that this interface and two classes be made private, since they aren't really a part of the core features that need to be provided by this package, and they clutter the namespace.

Also, this docstring in this position doesn't attach to the interface and classes - it will either attach to the package or get discarded (I don't remember which one) - neither is what you want, so this docstring should either be turned into a comment, or moved into the interface and class defs.